aspects/aspects_test.go

```diff
@@ -1270,3 +1270,203 @@ func (s *aspectSuite) TestSetOverwriteValueWithNewLevel(c *C) {
 	c.Assert(err, IsNil)
 	c.Assert(data, DeepEquals, map[string]interface{}{"bar": "baz"})
 }
+
+func (s *aspectSuite) TestSetValidatesDataWithSchemaPass(c *C) {
+	schema, err := aspects.ParseSchema([]byte(`{
+	"types": {
+		"int-map": {
+			"type": "map",
+			"values": {
+				"type": "int",
+				"min": 0
+			}
+		},
+		"str-array": {
+			"type": "array",
+			"values": {
+				"type": "string"
+			}
+		}
+	},
+	"schema": {
+		"foo": "$int-map",
+		"bar": "$str-array"
+	}
+}`))
+	c.Assert(err, IsNil)
+
+	databag := aspects.NewJSONDataBag()
+	aspectBundle, err := aspects.NewAspectBundle("acc", "bundle", map[string]interface{}{
+		"foo": []map[string]string{
+			{"request": "foo", "storage": "foo"},
+			{"request": "bar", "storage": "bar"},
+		},
+	}, schema)
+	c.Assert(err, IsNil)
+
+	asp := aspectBundle.Aspect("foo")
+	c.Assert(asp, NotNil)
+
+	err = asp.Set(databag, "foo", map[string]int{"a": 1, "b": 2})
+	c.Assert(err, IsNil)
+
+	err = asp.Set(databag, "bar", []string{"one", "two"})
+	c.Assert(err, IsNil)
+}
+
+func (s *aspectSuite) TestSetPreCheckValueFailsIncompatibleTypes(c *C) {
+	type schemaType struct {
+		schemaStr string
+		typ       string
+		value     interface{}
+	}
+
+	types := []schemaType{
+		{
+			schemaStr: `"int"`,
+			typ:       "int",
+			value:     int(0),
+		},
+		{
+			schemaStr: `"number"`,
+			typ:       "number",
+			value:     float64(0),
+		},
+		{
+			schemaStr: `"string"`,
+			typ:       "string",
+			value:     "foo",
+		},
+		{
+			schemaStr: `"bool"`,
+			typ:       "bool",
+			value:     true,
+		},
+		{
+			schemaStr: `{"type": "array", "values": "any"}`,
+			typ:       "array",
+			value:     []string{"foo"},
+		},
+		{
+			schemaStr: `{"type": "map", "values": "any"}`,
+			typ:       "map",
+			value:     map[string]string{"foo": "bar"},
+		},
+	}
+
+	for _, one := range types {
+		for _, other := range types {
+			if one.typ == other.typ || (one.typ == "int" && other.typ == "number") ||
+				(one.typ == "number" && other.typ == "int") {
+				continue
+			}
+
+			schema, err := aspects.ParseSchema([]byte(fmt.Sprintf(`{
+	"schema": {
+		"foo": %s,
+		"bar": %s
+	}
+}`, one.schemaStr, other.schemaStr)))
+			c.Assert(err, IsNil)
+
+			databag := aspects.NewJSONDataBag()
+			aspectBundle, err := aspects.NewAspectBundle("acc", "bundle", map[string]interface{}{
+				"foo": []map[string]string{
+					{"request": "foo", "storage": "foo", "access": "write"},
+					{"request": "foo", "storage": "bar", "access": "write"},
+				},
+			}, schema)
+			c.Assert(err, IsNil)
+
+			asp := aspectBundle.Aspect("foo")
+			c.Assert(asp, NotNil)
+
+			err = asp.Set(databag, "foo", one.value)
+			c.Assert(err, ErrorMatches, fmt.Sprintf(`storage paths "bar" and "foo" require incompatible types: %s != %s`, other.typ, one.typ))
```

Comment:
> Maybe `storage paths %q and %q for request %q require incompatible types`?
Will add the request to the error. Thanks

> Also, is there a reason that you couldn't check for these incompatible types in `aspects.NewAspectBundle`?

As I mentioned in the PR description, this can/will be used in the future to validate the aspect-bundle assertion. We just don't have that code right now